# [2장 2강] 실습: 선형회귀 모델 학습 방법

## 실습 목표

- 회귀 모델의 예측 오차를 MAE와 RMSE로 평가할 수 있다.
- NumPy 최소제곱법으로 선형회귀의 Weight와 Bias를 계산할 수 있다.
- scikit-learn으로 다중 선형회귀 모델을 학습할 수 있다.
- PyTorch와 Gradient Descent로 선형회귀 모델을 반복 학습할 수 있다.
- 세 가지 구현 방법의 파라미터와 성능을 비교할 수 있다.

## 진행 방식

- Google Colab 또는 Jupyter Notebook에서 진행합니다.
- 각 문제의 설명과 요구사항을 확인한 뒤 코드 셀을 작성합니다.
- 실행 결과를 확인하고 문제에 제시된 질문에 답합니다.

## 사용 데이터

- 파일: `insurance(2).csv`
- 입력 Feature: `age`, `sex`, `bmi`, `children`, `smoker`, `region`
- 예측 대상 Label: `charges`

## 실습 준비: 라이브러리 및 데이터 불러오기

Google Colab에서는 파일을 Google Drive의 `MyDrive`에 저장합니다. Jupyter Notebook에서는 노트북과 같은 폴더에 저장합니다.

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

try:
    from google.colab import drive
    drive.mount('/content/drive')
    file_path = '/content/drive/MyDrive/insurance.csv'
except ModuleNotFoundError:
    file_path = 'insurance.csv'

insurance_df = pd.read_csv(file_path)
display(insurance_df.head())
print('데이터 크기:', insurance_df.shape)
insurance_df.info()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


데이터 크기: (1338, 7)
<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 73.3 KB


## 필수 1: 의료비 데이터 확인 및 학습 데이터 준비

### 문제 1-1: 데이터 확인과 전처리 수행하기

#### 문제 설명

보험사는 고객 특성을 이용해 의료비를 예측하려고 합니다. 모델 학습 전에 데이터 품질과 `charges` 분포를 확인하고, 문자형 Feature를 숫자로 변환한 뒤 학습 데이터와 테스트 데이터로 나눕니다.

#### 요구사항

1. 중복 행 수와 전체 결측치 수를 확인합니다.
2. 중복 행을 제거하고 데이터 크기를 출력합니다.
3. `charges`의 평균, 중앙값, 최솟값, 최댓값을 출력합니다.
4. `sex`, `smoker`, `region`에 첫 번째 범주를 제외한 One-Hot Encoding을 적용합니다.
5. 입력 데이터 `X`와 예측 대상 `y`를 분리합니다.
6. `random_state=42`를 사용하여 학습용 80%, 테스트용 20%로 분리합니다.
7. 수치형 입력 컬럼인 `age`, `bmi`, `children`에 StandardScaler를 적용하되, 학습 데이터에는 `fit_transform()`, 테스트 데이터에는 `transform()`을 사용합니다.
8. 학습 데이터와 테스트 데이터의 크기 및 입력 컬럼을 출력합니다.

#### 출력 결과

- 중복 행 수 `1`, 전체 결측치 수 `0`
- 중복 제거 후 데이터 크기 `(1337, 7)`
- `charges` 주요 통계값
- 인코딩된 입력 컬럼 8개
- 학습 데이터 `(1069, 8)`, 테스트 데이터 `(268, 8)`

#### 결과 해석 작성

> 테스트 데이터에 `fit_transform()`이 아니라 `transform()`만 사용하는 이유는 무엇인가요?

## 필수 2: scikit-learn 다중 선형회귀 모델 학습

### 문제 2-1: 다중 선형회귀 모델 학습 및 평가하기

#### 문제 설명

표준화된 여러 입력 Feature를 동시에 사용하는 다중 선형회귀 기준 모델을 학습하고, 테스트 데이터의 의료비를 예측한 뒤 MAE와 RMSE로 평가합니다.

#### 요구사항

1. `LinearRegression` 모델을 생성합니다.
2. 표준화된 학습 데이터로 모델을 학습합니다.
3. 테스트 데이터의 의료비를 예측합니다.
4. Bias와 각 입력 Feature의 Weight를 출력합니다.
5. MAE와 RMSE를 계산합니다.
6. 실제값과 예측값 앞 5개를 비교합니다.
7. MAE와 RMSE의 의미와 RMSE가 더 크게 나타나는 이유를 설명합니다.

#### 결과 해석 작성

> MAE와 RMSE는 각각 무엇을 의미하며, 이 결과에서 RMSE가 MAE보다 큰 이유는 무엇인가요?

## 필수 3: NumPy 최소제곱법으로 선형회귀 직접 구현

### 문제 3-1: 최소제곱법으로 Weight와 Bias 계산하기

#### 문제 설명

NumPy 행렬 연산과 최소제곱법을 이용해 Weight와 Bias를 직접 계산하고 scikit-learn 결과와 비교합니다.

#### 요구사항

1. Bias 계산을 위해 학습 데이터 앞에 값이 모두 `1`인 컬럼을 추가합니다.
2. `np.linalg.pinv()`를 사용하여 최소제곱법의 파라미터를 계산합니다.
3. 첫 번째 파라미터를 Bias, 나머지를 Weight로 구분합니다.
4. 테스트 데이터에도 값이 모두 `1`인 컬럼을 추가하고 의료비를 예측합니다.
5. MAE와 RMSE를 계산합니다.
6. scikit-learn 결과와 Bias, Weight, 성능이 비슷한지 확인합니다.

#### 결과 해석 작성

> NumPy와 scikit-learn의 결과가 거의 같은 이유는 무엇인가요?

## 필수 4: PyTorch 경사하강법으로 선형회귀 학습

### 문제 4-1: Gradient Descent로 모델 학습하기

#### 문제 설명

PyTorch로 선형회귀 모델을 만들고 SGD를 이용해 MSE Loss가 감소하도록 Weight와 Bias를 반복적으로 업데이트합니다.

#### 요구사항

1. 재현성을 위해 `torch.manual_seed(42)`를 설정합니다.
2. 학습·테스트 입력과 Label을 `float32` Tensor로 변환합니다.
3. `torch.nn.Linear`로 선형회귀 모델을 생성합니다.
4. MSELoss와 SGD Optimizer를 사용하고 Learning Rate를 `0.01`로 설정합니다.
5. 2,000 Epoch 동안 학습하고 첫 Epoch와 200 Epoch마다 Loss를 출력합니다.
6. 학습 전후 Loss를 비교합니다.
7. 테스트 데이터로 MAE와 RMSE를 계산하고 Bias와 Weight를 출력합니다.
8. Gradient Descent가 Loss를 줄이는 원리를 설명합니다.

#### 결과 해석 작성

> Epoch가 증가하면서 Loss가 감소하는 이유는 무엇이며, Gradient Descent는 어떤 방향으로 파라미터를 업데이트하나요?

## 심화 1: 세 가지 선형회귀 모델 비교

### 문제 5-1: 모델별 파라미터와 성능 비교하기

#### 문제 설명

동일한 데이터로 학습한 NumPy, scikit-learn, PyTorch 모델의 Bias, Weight와 성능을 표로 비교합니다.

#### 요구사항

1. 세 모델의 Bias, MAE, RMSE를 하나의 표로 만듭니다.
2. 각 입력 Feature의 Weight를 모델별로 비교합니다.
3. 세 모델의 결과가 비슷한지 확인합니다.
4. 계산 방법이 다른데도 결과가 비슷한 이유를 설명합니다.

#### 결과 해석 작성

> 최소제곱법과 Gradient Descent의 계산 과정이 다른데도 세 모델의 결과가 비슷하게 나타나는 이유는 무엇인가요?

## 실습 마무리

1. 최소제곱법과 Gradient Descent의 가장 큰 차이는 무엇인가요?
2. Learning Rate가 너무 크거나 너무 작으면 어떤 문제가 발생하나요?
3. 세 모델을 공정하게 비교하려면 어떤 조건을 같게 유지해야 하나요?